### ISSUE TITLE
Leviathan Turrets (Issues that affect all of them)

### ISSUE THREAD

**Issue by HellDragon-HK** [GreatEmerald/UT3Vehicles#118]
- ANIMATION RELATED
   - [ ] Turrets are not centered on their arm once deployed

**Comment by GreatEmerald** [comment]
Not sure what you mean by "centred on their arm".

I think the turret operators should die relatively easily, the shield is there so they could defend themselves. It's even better than the UT3 system, as it depends on how much HP the operators had when they entered the vehicle.

Not sure about the colour of the shields... I'd be fine either way.

**Comment by HellDragon-HK** [comment]
By centered on arm I mean the turrets not being centered on the extending part they sit on, they are centered fine until they become occupied and then are sitting off the edge a little after the deploy (turret not main cannon) animation, I've tried seeing if translation could fix this but nope and the UT3 drawscale also did not fix it

Bots on my game are not even able to use a turret before a tank shell obliterates them that I've seen but I'll keep watch more, come to think of it I don't think I've seen them use a shield either....well I've got videos so I'll watch that first and see if I'm wrong

I like the shock color shield but purple is my favorite color so I'm very biased on it, team shields are good too but they are hard to see (one reason I'm having trouble getting the size right, I just can't see where the shield ends.....probably not an issue if I wasn't trying on RedPlanet however...stupid me)

**Comment by HellDragon-HK** [comment]
On a network game turret operators (maybe just bots) don't seem to take any damage

**Comment by HellDragon-HK** [comment]
New Issues:
Turret Operators are not taking damage in both online & offline games
Shield Sound is coming from somewhere other than the turrets, the beam turret it can be easily, the rocket turret less so, and the back turrets not at all

**Comment by GreatEmerald** [comment]
Oh yea, now I remember the arm alignment issue. I think it has to do with the bone, it extends further from the arms than it should. @100GPing100 Do you recall why that would be the case?

Well, tanks are pretty good at taking out Leviathan passengers, that's good balancing :)

And yes, I recall the issues with shields not always having collision for some reason.

Hm, good to know about sound origin, an incorrect variable must be used for the location...

**Comment by GreatEmerald** [comment]
Now I recall what the issue was with shield collision. IIRC in UT3 the passenger can shoot through the shield, but it stops incoming projectiles. It doesn't seem possible to do so in UT2004, because the shield either collides with projectiles and/or hitscan weapons or it doesn't.

I can think of two ways to work around the issue. One would be to spawn projectiles/hitscan beyond where the shield is, so what you're shooting never hits the shield. Another is to do something akin to what the stock Paladin does: allow shooting the shield, and if it's your own shot, have the shield not take damage but rather shoot the projectile out itself. This is more complicated, though, especially to make it work online.

**Comment by HellDragon-HK** [comment]
Would it help to move the shields closer so the gun barrels are through them?

Also why do the shields not block anything, the drawscale is set, the prepivot is set what else could there be, I've even tried changing the extends class from emitter to UT3PaladinShield & ONSShockTankShield, so far nothing works.

**Comment by GreatEmerald** [comment]
From what I can tell the collision seems fine. I'm guessing the invisible mesh is incorrectly placed. Set `bHidden=False` to be able to see where it is.

**Comment by HellDragon-HK** [comment]
I see the shields now, they are slightly off on the Z but shouldn't be enough to not block anything, in fact they should have blocked the Raptor plasma easily....I'll try to test it out, could take a bit

**Comment by HellDragon-HK** [comment]
Now it's actually working and I can't shoot through it, as soon as I set it back to true on being hidden it doesn't work though, any idea why?

**Comment by GreatEmerald** [comment]
What. That doesn't make any sense, the Paladin shield is set to hidden and it has collision fine... What if you attach the Paladin shield (replace all references to the Leviathan shield with references to the Paladin shield), does it have collision then?

**Comment by HellDragon-HK** [comment]
Replace what exactly, StaticMesh, extends class,....other?

**Comment by GreatEmerald** [comment]
All instances of the word `UT3LeviathanShield` with `UT3PaladinShield`.

**Comment by HellDragon-HK** [comment]
In which class, UT3LeviathanShield or the UT3LeviathanTurret and change it to Paladin?

**Comment by GreatEmerald** [comment]
`UT3LeviathanTurret` (and `UT3Leviathan`, etc.)

**Comment by HellDragon-HK** [comment]
Oh boy this is a little complicated.....

C:\Unreal Anthology\UT2004\UT3Vehicles\Classes\UT3LeviathanTurretWeapon.uc(77) : Error, Type mismatch in '='
Compile aborted due to errors.

**Comment by GreatEmerald** [comment]
Please put ``` before and after the code blocks when pasting it here!

But when I said every instance, I ment every single instance of that word. `Shield` must be defined as `var UT3PaladinShield Shield;` too.

**Comment by HellDragon-HK** [comment]
Ah heck the shield is along the front of the wheels, I can't tell if it's the shield working or the Levi's collision...ugh I don't want to mess with the Paladin's shield placement to test that....going to have to though aren't I?

**Comment by HellDragon-HK** [comment]
Ok after adjusting the placement the shield does appear to work now but has a big problem, it doesn't move with the turret and seems to be connected to the Levi itself, I'll play with bots here in a second to test if it moves with the Levi or not

**Comment by HellDragon-HK** [comment]
Well this is interesting, not only does it not attach to the turret but it doesn't attach to anything, it just spawns and stays where the Levi's spawn point was, even when it deactivates and I reactivate it it's still there and it's also exactly where the sound is so the sound not moving with the shield might be one and the same issue

**Comment by GreatEmerald** [comment]
Hm, it should be attached by this: https://github.com/GreatEmerald/UT3Vehicles/blob/dbeef984f95699109dca8cb1ddf39da9225b3056/Classes/UT3LeviathanTurretWeapon.uc#L81
Which takes the bone name from: https://github.com/GreatEmerald/UT3Vehicles/blob/dbeef984f95699109dca8cb1ddf39da9225b3056/Classes/UT3LeviathanTurretWeaponBeam.uc#L164 (and other UT3LeviathanTurretWeapon* classes)
That said, `name`s are weird. I don't remember if any special handling is needed for them when setting them in defaultproperties. So check whether that bone exists, and check with the console whether `get UT3LeviathanTurretWeaponBeam ShieldAttachmentBone` gives you the right name.

**Comment by HellDragon-HK** [comment]
Yes that bone exists, it's what the camera was originally set on and it's the `WeaponFireAttachmentBone`, not sure about in console but something I have to ask, I've noticed the Leviathan vehicle strings are not working, I don't see 'in a Leviathan' or 'in a Leviathan insert turret type' might that also be related?

The console command confirmed Lt_Front_Turret_Barrel as well.

**Comment by GreatEmerald** [comment]
I get the feeling that you may be talking about the shield effect, rather than the hidden collision (the effect wouldn't get attached properly because it's set to Paladin values in `UT3PaladinShield). Did you set `bHidden=False` in `UT3PaladinShield`?

**Comment by HellDragon-HK** [comment]
I moved both the shield and the visual, it does have collision, I didn't set it to not hidden because I didn't think it was needed, also it needs to be hidden be default so if it's not working then it's a problem.

Also moving out a bit the scoreboard did change to 'in a Leviathan' but not a UT3 Leviathan like the string has it.....is this not working right or is it just being truncated you think? (ONS-RedPlanet mid field)

**Comment by GreatEmerald** [comment]
Well, make it not hidden and check where it actually is.

**Comment by HellDragon-HK** [comment]
Well I made it not hidden and....I have no idea if that helped or complicated things even further.....
![ss_001](https://user-images.githubusercontent.com/34013048/34653899-f692996a-f3b8-11e7-8782-782467bfd57b.jpg)
![ss_002](https://user-images.githubusercontent.com/34013048/34653906-0c6e1444-f3b9-11e7-8d80-d48c6b903fb5.jpg)

With it hidden

![ss_005](https://user-images.githubusercontent.com/34013048/34653907-2222935a-f3b9-11e7-900f-d361efa1a134.jpg)
![ss_008](https://user-images.githubusercontent.com/34013048/34653912-34c8a33c-f3b9-11e7-8c53-6bd8f6e47785.jpg)

With it not hidden the actual shield part moves and works like it should but the visual effect that appears is not attached and just shows up at the same location but doesn't move with....could be the difference in how the Paladin is a continuous shield where as the Leviathan is a timed shield perhaps?

With it hidden as you can see the shield and visual appear to be in the same place due the collision working but the shield always spawns at the Levi's spawn point....why does changing bHidden cause it to change so much?

**Comment by GreatEmerald** [comment]
All right, so we do know that it's attached properly, then. The fact that you can see collision happening at the start when it's hidden is rather odd though. But the collision may still be there when it's visible, just that you can't see it due to your shots hitting the shield. I wonder if `show collision` would give some clues.

And really, nobody replaced the shield mesh from the stock Paladin's... That needs to be fixed at some point too.

**Comment by HellDragon-HK** [comment]
Collision works in both hidden and non hidden, it's just the visual doesn't stay with the shield when not hidden and the entire shield doesn't stay with the vehicle when hidden.  Show Collision doesn't work on Paladin so probably not here either.

I didn't know about it until now, so I'll look at and see about it, I get the feeling doing that and we'll have the same non functional shield issue
